### PR TITLE
Implement resample of an array as Stream

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

File: src/Streamly/Statistics.hs [hunk 1]
```diff
@@ -736,3 +747,43 @@ binBoundaries = undefined
...
+    where
+
+    inject arr = do
+        liftIO $ return (arr, length arr, 0)
+
+    step (arr, len, idx) = liftIO $ do
+        if idx >= len
+        then return Stop
+        else do
+            g <- createSystemRandom
```
Comment (harendra-kumar) [inline src/Streamly/Statistics.hs:770]: `g` should be generated only once in `inject`.

File: src/Streamly/Statistics.hs [hunk 2]
```diff
@@ -736,3 +747,43 @@ binBoundaries = undefined
...
+
+{-# INLINE resampleOne #-}
+resampleOne :: (MonadIO m, Storable a) =>
+    Unfold m (Array a) a
+resampleOne = Unfold step inject
+
+    where
+
+    inject arr = do
+        liftIO $ return (arr, length arr, 0)
```
Comment (harendra-kumar) [inline src/Streamly/Statistics.hs:764]: We can use just `(arr, count)` where count = length arr. Keep decrementing the count and stop when it reaches 0.

File: src/Streamly/Statistics.hs [hunk 3]
```diff
@@ -736,3 +747,43 @@ binBoundaries = undefined
...
+    -> Fold m a b   -- ^ Estimation fold
+    -> SerialT m b
+resample n arr fld = do
+    let loop k strm
+            | k >= n = strm
+            | otherwise = do
+                let sampled = Stream.unfold resampleOne arr
+                let strm' = Stream.consM (Stream.fold fld sampled) strm
+                loop (k + 1) strm'
+        in loop 0 Stream.nil
```
Comment (harendra-kumar) [inline src/Streamly/Statistics.hs:789]: This looks ugly. We should not be re-implementing from scratch the combinators that we already have. We can just use `Stream.replicateM`.

File: test/Main.hs [hunk 4]
```diff
@@ -3,12 +3,18 @@
...
 
+import qualified Data.Set as Set
+import qualified Streamly.Data.Fold as Fold
 import qualified Streamly.Internal.Data.Ring.Foreign as Ring
 import qualified Streamly.Prelude as S
 
 import Prelude hiding (sum, maximum, minimum)
 
+sampleArray :: [Double]
+sampleArray = [1.0, 2.0, 3.0, 4.0, 5.0]
```
Comment (harendra-kumar) [inline test/Main.hs:16]: calling it "list" instead of array would be better.

File: test/Main.hs [hunk 5]
```diff
@@ -71,8 +77,28 @@ main = hspec $ do
...
+                ls <- runIO $ S.toList a
+                let s' = Set.fromList ls
+                    sub = Set.isSubsetOf s' s
+                it ("resampleOne " ++ show ls)
+                    (Prelude.length ls == Array.length c && sub)
 
+            testFuncResample n tc = do
+                let arr = Array.fromList tc
+                a <- runIO $ S.toList $ resample n arr Fold.mean
+                it ("resample" ++ show a) (Prelude.length a == n)
```
Comment (harendra-kumar) [inline test/Main.hs:93]: Give a white space after "resample".

**Review by harendra-kumar — CHANGES_REQUESTED** [review]
1) performance is interesting because of random number generation. Can you add a benchmark?
2) rebase

File: src/Streamly/Statistics.hs (hunk 2 above)
Comment (rnjtranjan) [reply src/Streamly/Statistics.hs:764]: we need the length unchanged for uniformRM (0, len - 1) g

**Pull request merged**